tests/Agency/StoreTestAPI.cpp

```diff
@@ -0,0 +1,1239 @@
+      ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+      res = write(R"([[{"/b":3},{"/a/e":{"in":3},"/a/b/c":{"in":3}}]])");
+      ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+      res = write(R"([[{"/b":3},{"/a/b/c":{"in":3},"/a/e":{"in":3}}]])");
+      ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+      res = write(R"([[{"/b":3},{"/a/b/c":{"in":3},"/a/e":{"in":2}}]])");
+      ASSERT_EQ(consensus::apply_ret_t::APPLIED, res.front());
+      assertEqual(readAndCheck(R"([["/b"]])"), R"([{"b":3}])");
+      
+      // Permute order of keys and objects within precondition
+      std::map<std::string,std::string> baz {
+        {"_id", "\"5a00203e4b660989b2ae5493\""},
+        {"index", "0"},
+        {"guid", "\"7a709cc2-1479-4079-a0a3-009cbe5674f4\""},
+        {"isActive", "true"},
+        {"balance", "\"$3,072.23\""},
+        {"picture", "\"http://placehold.it/32x32\""},
+        {"age", "21"},
+        {"eyeColor", "\"green\""},
+        {"name", R"({ "first": "Durham", "last": "Duke" })"},
+        {"tags", R"(["anim","et","id","do","est",1.0,-1024,1024])"}
+      };
+      bool is_first{true};
+      std::string const baz_text = "{" + 
+        std::accumulate(baz.begin(), baz.end(), std::string(), [&is_first](std::string partial, auto const &kv){
+          if (is_first) is_first = false; else partial += ",";
+          partial += "\"" + kv.first + "\": " + kv.second;
+          return partial;
+        })
+        + "}";
+      std::string const qux {R"(["3.14159265359",3.14159265359])"};
+      std::string const foo_value {"\"bar\""};
+      std::string const localObj = R"(
+          {"foo" : )"+ foo_value + R"(,
+           "baz" : )" + baz_text + R"(,
+           "qux" : )" + qux + R"(
+          })";
+
+      res = write(
+        std::string("[[") + localObj + R"(,
+          {
+           "baz":{"old": )" + baz_text + R"(},
+           "qux":)" + qux + R"(}]])");
+      ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+
+      writeAndCheck("[[" + localObj + "]]");
+      writeAndCheck(
+        "[[" + localObj + R"(, {"foo":)" + foo_value + R"(,"baz":{"old":)" + baz_text + R"(},"qux":)" + qux + "}]]");
+      writeAndCheck(
+        "[[" + localObj + R"(, {"baz":{"old":)" + baz_text + R"(},"foo":)" + foo_value + R"(,"qux":)" + qux + R"(}]])");
+      writeAndCheck(
+        "[[" + localObj + R"(, {"baz":{"old":)" + baz_text + R"(},"qux":)" + qux + R"(,"foo":)" + foo_value + R"(}]])");
+      writeAndCheck(
+        "[[" + localObj + R"(, {"qux":)" + qux + R"(,"baz":{"old":)" + baz_text + R"(},"foo":)" + foo_value + R"(}]])");
+
+      std::vector<std::string> localKeys;
+      std::transform(baz.begin(), baz.end(), std::back_inserter(localKeys), [](auto kv){ return kv.first; });
+      for (int permutation_count{}; permutation_count < 2000; ++permutation_count) {
```

Comment:
2000 is way too much and not necessary. 3 permutation will be more than enough. The code now seems to work.